pkg/migration/controllers/migrationschedule.go

```diff
@@ -0,0 +1,306 @@
+		err = m.pruneMigrations(migrationSchedule)
+		if err != nil {
+			msg := fmt.Sprintf("Error pruning old migrations: %v", err)
+			m.Recorder.Event(migrationSchedule,
+				v1.EventTypeWarning,
+				string(stork_api.MigrationStatusFailed),
+				msg)
+			log.MigrationScheduleLog(migrationSchedule).Error(msg)
+			return err
+		}
+	}
+	return nil
+}
+
+func (m *MigrationScheduleController) updateMigrationStatus(migrationSchedule *stork_api.MigrationSchedule) error {
+	updated := false
+	for _, policyMigration := range migrationSchedule.Status.Items {
+		for _, migration := range policyMigration {
+			// Get the updated status it we see it as not completed
+			if !m.isMigrationComplete(migration.Status) {
+				pendingMigration, err := k8s.Instance().GetMigration(migration.Name, migrationSchedule.Namespace)
+				if err != nil {
+					m.Recorder.Event(migrationSchedule,
+						v1.EventTypeWarning,
+						string(stork_api.MigrationStatusFailed),
+						fmt.Sprintf("Error getting status of migration %v: %v", migration.Name, err))
+					pendingMigration.Status.Status = stork_api.MigrationStatusFailed
+				}
+
+				if pendingMigration.Status.Status == stork_api.MigrationStatusInitial {
+					pendingMigration.Status.Status = stork_api.MigrationStatusPending
+				}
+
+				// Check again and update the status if it is completed
+				migration.Status = pendingMigration.Status.Status
+				if m.isMigrationComplete(migration.Status) {
+					migration.FinishTimestamp = meta.Now()
+					if pendingMigration.Status.Status == stork_api.MigrationStatusSuccessful {
+						m.Recorder.Event(migrationSchedule,
+							v1.EventTypeNormal,
+							string(stork_api.MigrationStatusSuccessful),
+							fmt.Sprintf("Scheduled migration (%v) completed successfully", migration.Name))
+					} else {
+						m.Recorder.Event(migrationSchedule,
+							v1.EventTypeWarning,
+							string(stork_api.MigrationStatusFailed),
+							fmt.Sprintf("Scheduled migration (%v) failed", migration.Name))
+					}
+				}
+				updated = true
+			}
+		}
+	}
+	if updated {
+		err := sdk.Update(migrationSchedule)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (m *MigrationScheduleController) isMigrationComplete(status stork_api.MigrationStatusType) bool {
+	if status == stork_api.MigrationStatusPending ||
+		status == stork_api.MigrationStatusCaptured ||
+		status == stork_api.MigrationStatusInProgress {
+		return false
+	}
+	return true
+}
+
+func (m *MigrationScheduleController) shouldStartMigration(migrationSchedule *stork_api.MigrationSchedule) (stork_api.SchedulePolicyType, bool, error) {
+	// Don't trigger a new migration if one is already in progress
+	for _, policyType := range stork_api.GetValidSchedulePolicyTypes() {
+		policyMigration, present := migrationSchedule.Status.Items[policyType]
+		if present {
+			for _, migration := range policyMigration {
+				if !m.isMigrationComplete(migration.Status) {
+					return stork_api.SchedulePolicyTypeInvalid, false, nil
```

Comment:
Why should this return SchedulePolicyTypeInvalid?

We found that a migration is not yet complete for `policyType`. So it should be

```
return policyType, false, nil
```

Reply:
The purpose of the function is to check if a migration should be started. If yes, then it should also return the policy type that needs to be triggered. It doesn't return the type of policy that has been triggered. I'll add comments to the function to make it clear.